src/include/duckdb/common/single_thread_pointer.h

```diff
@@ -0,0 +1,169 @@
+#include <cstdlib>
+
+template <typename T>
+class single_thread_ptr {
+public:
+	T *ptr;              // contained pointer
+	uint32_t *ref_count; // reference counter
+
+public:
+	// Default constructor, constructs an empty single_thread_ptr.
+	constexpr single_thread_ptr() : ptr(nullptr), ref_count(nullptr) {
+	}
+	// Construct empty single_thread_ptr.
+	constexpr single_thread_ptr(std::nullptr_t) : ptr(nullptr), ref_count(nullptr) {
+	}
+	// Construct a single_thread_ptr that wraps raw pointer.
+	explicit single_thread_ptr(T *p) {
+		if (p) {
+			ptr = p;
+			ref_count = new uint32_t(1);
```

Comment:
Another question; in the `release` it seems that only `ptr` is deleted. That is correct for pointers created using `make_shared`, but it appears to leak memory if this constructor is used, since `ref_count` would never be freed.

Is this constructor used anywhere? If not, perhaps remove it entirely? Or add a "owned_ref_count" flag so this can be correctly freed?

Reply:
No, I will remove them, thanks